pvlib/modelchain.py

```diff
@@ -1238,29 +1514,50 @@ def _prepare_temperature(self, data=None):
         -------
         self
 
-        Assigns attribute ``cell_temperature``.
+        Assigns attribute ``results.cell_temperature``.
 
         """
-        if 'cell_temperature' in data:
-            self.cell_temperature = data['cell_temperature']
+        if not isinstance(data, tuple) and self.system.num_arrays > 1:
+            data = (data,) * self.system.num_arrays
+        elif not isinstance(data, tuple):
+            return self._prepare_temperature_single_array(data)
+        given_cell_temperature = tuple(itertools.starmap(
+            self._get_cell_temperature,
+            zip(data, self.results.total_irrad,
+                self.system.temperature_model_parameters)
+        ))
+        # If cell temperature has been specified for all arrays return
+        # immediately and do not try to compute it.
+        if all(cell_temp is not None for cell_temp in given_cell_temperature):
+            self.results.cell_temperature = given_cell_temperature
             return self
-
-        # cell_temperature is not in input. Calculate cell_temperature using
-        # a temperature_model.
-        # If module_temperature is in input data we can use the SAPM cell
-        # temperature model.
-        if (('module_temperature' in data) and
-                (self.temperature_model.__name__ == 'sapm_temp')):
-            # use SAPM cell temperature model only
-            self.cell_temperature = pvlib.temperature.sapm_cell_from_module(
-                module_temperature=data['module_temperature'],
-                poa_global=self.total_irrad['poa_global'],
-                deltaT=self.system.temperature_model_parameters['deltaT'])
-            return self
-
-        # Calculate cell temperature from weather data. Cell temperature models
-        # expect total_irrad['poa_global'].
+        # Calculate cell temperature from weather data. If cell_temperature
+        # has not been provided for some arrays then it is computed with
+        # ModelChain.temperature_model(). Because this operates on all Arrays
+        # simultaneously, 'poa_global' must be known for all arrays, including
+        # those that have a known cell temperature.
+        try:
+            self._verify_df(self.results.total_irrad, ['poa_global'])
+        except ValueError:
+            # Provide a more informative error message. Because only
+            # run_model_from_effective_irradiance() can get to this point
+            # without known POA we can suggest a very specific remedy in the
+            # error message.
+            raise ValueError("Incomplete input data. Data must contain "
```

Comment:
@wfvining @wholmgren here is where we would need to substitute `effective_irradiance` for `poa_global` for the cell temperature models. We [discussed](https://github.com/pvlib/pvlib-python/pull/943#discussion_r475890038) doing so as part of #943 but decided against it.

Reply:
I'd prefer a follow up PR. This one is already enormous and hard to review.

Reply:
#1120